modules/dnn/src/onnx/onnx_graph_simplifier.cpp

```diff
@@ -134,6 +134,238 @@ class ONNXGraphWrapper : public ImportGraphWrapper
+            opencv_onnx::NodeProto* constant_node = constant_ptr.dynamicCast<ONNXNodeWrapper>()->node;
+            opencv_onnx::TensorProto constant_proto = constant_node->attribute(0).t();
+            Mat constant_mat = getMatFromTensor(constant_proto);
+            return *constant_mat.ptr<T>(return_mat_ele_idx);
+        }
+    }
+
+    static void getAttributes(const Ptr<ImportGraphWrapper>& net, std::vector<int>& matchedNodesIds, int64_t& _num_heads, std::vector<int64_t>& _qkv_hidden_sizes, float& _scale) {
+        _qkv_hidden_sizes.clear();
+        int64_t qk_hidden_size = 0;
+        for (size_t i = 0; i < matchedNodesIds.size(); i++) {
+            auto id = matchedNodesIds[i];
+            const auto node = net->getNode(id);
+            if (node->getType() == "Slice") { // collect hidden_sizes
+                int slice_start = extractConstant<int>(net, id, 1, 0);
+                int slice_end = extractConstant<int>(net, id, 2, 0);
+                int64_t hidden_size = static_cast<int64_t>(slice_end - slice_start);
+                if (slice_start == 0) {
+                    qk_hidden_size = hidden_size;
+                }
+                _qkv_hidden_sizes.push_back(hidden_size);
+            } else if (node->getType() == "Reshape" && i != matchedNodesIds.size() - 1) { // collect num_heads
+                _num_heads = extractConstant<int>(net, id, 1, 1);
+            } else if (node->getType() == "Div") { // collect scale
+                _scale = extractConstant<float>(net, id, 1, 0);
+            }
+        }
+
+        CV_CheckGT(qk_hidden_size, 0, "Attention fusion: failed to get qk_hidden_size");
+        CV_CheckEQ(_qkv_hidden_sizes.size(), static_cast<size_t>(3), "Attention fusion: invalid qkv hidden sizes");
+
+        int64_t v_hidden_size = 0;
+        for (auto h: _qkv_hidden_sizes) {
+            if (h != qk_hidden_size) {
+                v_hidden_size = h;
+            }
+        }
+        if (v_hidden_size == 0) {
+            v_hidden_size = qk_hidden_size;
+        }
+        _qkv_hidden_sizes[0] = qk_hidden_size;
+        _qkv_hidden_sizes[1] = qk_hidden_size;
+        _qkv_hidden_sizes[2] = v_hidden_size;
+    }
+
+    virtual bool match(const Ptr<ImportGraphWrapper>& net, int nodeId,
+                       std::vector<int>& matchedNodesIds,
+                       std::vector<int>& targetNodesIds) CV_OVERRIDE {
+        if (Subgraph::match(net, nodeId, matchedNodesIds, targetNodesIds))
+        {
+            // get weight
+            weight_name = getInputName(net, matchedNodesIds[0], 1);
+            // std::cout << "attention: weight_name=" << weight_name << std::endl;
+
+            // get bias
+            bias_name = getInputName(net, matchedNodesIds[1], 0);
+            // std::cout << "attention: bias_name=" << bias_name << std::endl;
+
+            // get attrs
+            getAttributes(net, matchedNodesIds, num_heads, qkv_hidden_sizes, scale);
```

Comment:
Why not a part of `finalize`?

Reply:
If it is put in `finalize`, member variables are not needed then. It should also work.

Reply:
Just realize that `getAttributes` needs `matchedNodesIds` which is not existed in `finalize`.